src/torch_models.py

```diff
@@ -146,6 +146,105 @@ def _fit(self, inputs: Array, outputs: Array) -> None:
+        self._output_scale = np.zeros(1, dtype=np.float32)
+        self._classifier = MLPClassifier(1, 1)
+
+    def fit(self, X: Array, Y: Array) -> None:
+        # Normalize everything right off the bat for simplicity.
+        X, self._input_shift, self._input_scale = self._normalize_data(X)
+        Y, self._output_shift, self._output_scale = self._normalize_data(Y)
+        # Initialize the classifier.
+        num_data, self._x_dim = X.shape
+        assert Y.shape[0] == num_data
+        logging.info(f"Training {self.__class__.__name__} on {num_data} "
+                     "datapoints")
+        self._y_dim = Y.shape[1]
+        max_itr = CFG.implicit_mlp_regressor_max_itr
+        self._classifier = MLPClassifier(in_size=(self._x_dim + self._y_dim),
+                                         max_itr=max_itr,
+                                         balance_data=False)
+        # Create the negative data.
+        neg_X, neg_Y = self._create_negative_data(X, Y)
+        # Set up the data for the classifier.
+        pos_concat_inputs = np.hstack([X, Y])
+        neg_concat_inputs = np.hstack([neg_X, neg_Y])
+        concat_inputs = np.vstack([pos_concat_inputs, neg_concat_inputs])
+        targets = np.array([1 for _ in pos_concat_inputs] +
+                           [0 for _ in neg_concat_inputs],
+                           dtype=np.float32)
+        # Train the classifier.
+        self._classifier.fit(concat_inputs, targets)
+
+    def _create_negative_data(self, X: Array, Y: Array) -> Tuple[Array, Array]:
+        """This makes the assumption that negative data are far, far more
+        common than positive data.
+
+        Under this assumption, the negative data are simply randomly
+        sampled. There may be false negatives in general, but they
+        should be rare, under the assumption.
+        """
+        # Note that the data has already been normalized.
+        del Y  # unused for now, but may be used in the future.
+        num_samples = CFG.implicit_mlp_regressor_num_negative_data_per_input
+        neg_input_lst = []
+        neg_output_lst = []
+        for pos_input in X:
+            samples = self._rng.uniform(size=(num_samples, self._y_dim))
+            for neg_output in samples:
+                neg_input_lst.append(pos_input)
+                neg_output_lst.append(neg_output)
+        neg_inputs = np.array(neg_input_lst, dtype=np.float32)
+        neg_outputs = np.array(neg_output_lst, dtype=np.float32)
+        return neg_inputs, neg_outputs
+
+    def predict(self, arr: Array) -> Array:
+        # This sampling-based inference method is okay in 1 dimension, but
+        # won't work well with higher dimensions.
+        assert arr.shape == (self._x_dim, )
+        # Normalize.
+        x = (arr - self._input_shift) / self._input_scale
+        num_samples = CFG.implicit_mlp_regressor_num_samples_per_inference
+        sample_ys = self._rng.uniform(size=(num_samples, self._y_dim))
```

Comment:
how does it make sense to do this uniformly? what if certain object states are nonsensical or infeasible to reach?

Reply:
well, they're negative data, so we would hope to not strive toward the nonsensical or infeasible things. but you're right, it feels like there is something better that can be done here, using the distribution seen in the data. since the implicit BC paper just does what we're doing here, I'm going to leave it for now.